issue #1315 - instantiates field represents the registered capabilities

tbieste opened the pull request:
This change enables the CapabilityStatement.instantiates field returned by the metadata operation to return the URIs of the CapabilityStatements registered with the FHIR server.

Signed-off-by: Troy Biesterfeld <tbieste@us.ibm.com>

prb112 commented on operation/fhir-operation-bulkdata/src/main/resources/hl7/fhir/uv/bulkdata/package/CapabilityStatement-bulk-data.json:
```diff
@@ -0,0 +1 @@
+{"resourceType":"CapabilityStatement","id":"bulk-data","text":{"status":"generated","div":"<div xmlns=\"http://www.w3.org/1999/xhtml\"><h2>BulkDataIGCapabilityStatement</h2><div><p>This Section describes the expected capabilities of a FHIR Bulk Data Server actor which is responsible for providing responses to the queries submitted by the FHIR Bulk Data Requestors. The complete list of FHIR profil ...
```
Note, we also support import, which is not listed here, and we've jammed into the capability statement.

tbieste replied:
Are you suggesting an update to the capability statement?  Only its URL is externalized via the metadata operation.

prb112 replied:
I'm just warning that we have one more operation in case we need to create a custom capability statement for this case.

prb112 replied:
I think $import is at the System level and Bulk Status is at the System Level, I think this doesn'timpact it.  Right?

tbieste replied:
Right, the bulkdata Capability Statement only mentions export.  Import is additional functionality not mentioned by the bulkdata Capability Statement.

prb112 replied:
Right - to clarify, I'm just making sure that it is still listed in the `metadata`

tbieste replied:
Yes, this does not change the values returned in the CapabilityStatement->rest->operation field, which is where the import operation is described.

prb112 replied:
great 👍

tbieste commented on fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:
```diff
@@ -266,25 +258,14 @@ private CapabilityStatement buildCapabilityStatement() throws Exception {
-                    .system(Uri.of("http://terminology.hl7.org/CodeSystem/restful-security-service"))
-                    .build())
-                .build());
+            securityBuilder.service(CodeableConcept.builder().coding(Coding.builder().code(Code.of("Certificates")).system(Uri.of("http://terminology.hl7.org/CodeSystem/restful-security-service")).build()).build());
```
Fixed.

prb112 reviewed (approved):
LGTM

lmsurpre reviewed (changes requested):
Please use the CapabilityStatement from http://build.fhir.org/ig/HL7/bulk-data/branches/tech-correct/CapabilityStatement-bulk-data.html instead of the currently published one...this branch is what should have been published last month.

tbieste commented:
> Please use the CapabilityStatement from http://build.fhir.org/ig/HL7/bulk-data/branches/tech-correct/CapabilityStatement-bulk-data.html instead of the currently published one...this branch is what should have been published last month.
> 
> See https://chat.fhir.org/#narrow/stream/179250-bulk-data/topic/capabilities.2Coperation.2Ereference.2Edefinition/near/211955165 for more detail.
I've updated to use the CapabilityStatement and OperationDefinitions from http://build.fhir.org/ig/HL7/bulk-data/branches/tech-correct/index.html.  This changes the version of the bulkdata operations from 1.0.0 to 1.0.1, so I also updated the version of the export-status and import operations from 1.0.0 to 1.0.1.

lmsurpre commented on fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:
```diff
@@ -91,6 +92,8 @@
     private static final String FHIR_SERVER_NAME = "IBM FHIR Server";
     private static final String FHIR_COPYRIGHT = "(C) Copyright IBM Corporation 2016, 2020";
     private static final String EXTENSION_URL = "http://ibm.com/fhir/extension";
+    private static final String BASE_CAPABILITY_URL = "http://hl7.org/fhir/CapabilityStatement/base";
+    private static final String BASE_2_CAPABILITY_URL = "http://hl7.org/fhir/CapabilityStatement/base2";
```
what are these for?

tbieste replied:
The FHIR registry includes these two capability statements:
/fhir-registry/src/main/resources/hl7/fhir/core/package/CapabilityStatement-base.json
/fhir-registry/src/main/resources/hl7/fhir/core/package/CapabilityStatement-base2.json
Since they seemed related to the base FHIR spec and have experimental = true, I wasn't sure they should be included in "instantiates".

lmsurpre replied:
interesting.  i agree we should exclude them.  i wonder whether they should truly be in the registry at all.
@JohnTimm thoughts?

lmsurpre replied:
if we don't here from john by the morning, I think we should just add a comment (either here or where I just suggested it below) and merge this one in time for 4.4.1

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/resources/OperationDefinition-group-export.json:
```diff
@@ -83,14 +83,6 @@
       "documentation": "A string of comma-delimited FHIR resource types. Only resources of the specified resource types(s) SHOULD be included in the response. If this parameter is omitted, the server SHOULD return all supported resources within the scope of the client authorization. For non-system-level requests, the Patient Compartment SHOULD be used as a point of reference for recommended resou ...
       "type": "string"
-    },
-    {
-      "name": "_typeFilter",
-      "use": "in",
-      "min": 0,
-      "max": "1",
-      "documentation": "A string of comma-delimited type filters.",
-      "type": "string"
```
i'm disappointed the removed these from the OperationDefinition.  I think we want to continue supporting these...we might want to add it back in :-(

tbieste replied:
Adding back in.

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/resources/OperationDefinition-patient-export.json:
```diff
@@ -83,14 +83,6 @@
       "documentation": "A string of comma-delimited FHIR resource types. Only resources of the specified resource types(s) SHOULD be included in the response. If this parameter is omitted, the server SHOULD return all supported resources within the scope of the client authorization. For non-system-level requests, the Patient Compartment SHOULD be used as a point of reference for recommended resou ...
       "type": "string"
-    },
-    {
-      "name": "_typeFilter",
-      "use": "in",
-      "min": 0,
-      "max": "1",
-      "documentation": "A string of comma-delimited type filters.",
-      "type": "string"
```
i'm disappointed the removed these from the OperationDefinition.  I think we want to continue supporting these...we might want to add it back in :-(

tbieste replied:
Adding back in.

lmsurpre commented on operation/fhir-operation-bulkdata/src/main/resources/export.json:
```diff
@@ -81,15 +81,6 @@
       "documentation": "A string of comma-delimited FHIR resource types. Only resources of the specified resource types(s) SHOULD be included in the response. If this parameter is omitted, the server SHOULD return all supported resources within the scope of the client authorization. Resource references MAY be relative URIs with the format <resource type>/<id>, or absolute URIs with the same struc ...
       "type": "string"
-    },
-    {
-      "name": "_typeFilter",
-      "use": "in",
-      "min": 0,
-      "max": "1",
-      "documentation": "A string of comma-delimited type filters.",
-      "type": "string"
-    }    
```
i'm disappointed the removed these from the OperationDefinition.  I think we want to continue supporting these...we might want to add it back in :-(

prb112 replied:
yes - let's add back in

tbieste replied:
Adding back in.

lmsurpre commented on fhir-server/src/main/java/com/ibm/fhir/server/resources/Capabilities.java:
```diff
@@ -380,6 +383,32 @@ private CapabilityStatement buildCapabilityStatement() throws Exception {
+        for (CapabilityStatement registeredCapability : registeredCapabilities) {
+            if (registeredCapability != null && registeredCapability.getUrl() != null) {
+                String url = registeredCapability.getUrl().getValue();
+                if (url != null && !BASE_CAPABILITY_URL.equals(url) && !BASE_2_CAPABILITY_URL.equals(url)) {
```
```suggestion
                // BASE_CAPABILITY_URL and BASE_2_CAPABILITY_URL come from the core spec and shouldn't be advertised
                if (url != null && !BASE_CAPABILITY_URL.equals(url) && !BASE_2_CAPABILITY_URL.equals(url)) {
```

lmsurpre reviewed (approved):
Left a single suggestion to add a comment.  Otherwise LGTM.
